File: scripts/python.py
"""Python adapter commands."""

import argparse
import os
import subprocess
import sys

from shared.constants import PYTHON_DIR
from shared.run import run
from shared.format import check

CMDS = {
    "setup": ["uv", "sync", "--extra", "dev", "--extra", "websocket"],
    "test": ["uv", "run", "pytest"],
}


def fmt_check() -> int:
    return check("Python",
                 ["uvx", "ruff", "format", "--check", "."],
                 ["uvx", "ruff", "format", "."],
                 cwd=PYTHON_DIR)


def lint() -> int:
    result = subprocess.run(
        ["uvx", "ruff", "check", "."], cwd=PYTHON_DIR, capture_output=True, text=True,
    )
    if result.returncode == 0:
        return 0
    print(result.stdout, result.stderr, sep="", end="", flush=True)
    if os.environ.get("CI"):
        return result.returncode
    subprocess.run(["uvx", "ruff", "check", ".", "--fix"], cwd=PYTHON_DIR)
    print("[WARN] Python lint issues auto-fixed. Stage changes before committing.", flush=True)
    return result.returncode


def clean() -> int:
    import shutil
    shutil.rmtree(PYTHON_DIR / ".venv", ignore_errors=True)
    shutil.rmtree(PYTHON_DIR / ".ruff_cache", ignore_errors=True)
    shutil.rmtree(PYTHON_DIR / ".pytest_cache", ignore_errors=True)
    for egg in PYTHON_DIR.glob("*.egg-info"):
        shutil.rmtree(egg, ignore_errors=True)
    for pycache in PYTHON_DIR.rglob("__pycache__"):
        shutil.rmtree(pycache, ignore_errors=True)
    for pyc in PYTHON_DIR.rglob("*.pyc"):
        pyc.unlink(missing_ok=True)
    return 0


def main() -> None:
    parser = argparse.ArgumentParser(description="Python adapter commands")
    parser.add_argument("command", nargs="?", default="check",
                        choices=["check", "fmt_check", "lint", "clean"] + list(CMDS))
    args = parser.parse_args()
    if args.command == "fmt_check":
        sys.exit(fmt_check())
    if args.command == "lint":
        sys.exit(lint())
    if args.command == "check":
        failed = any([fmt_check() != 0, lint() != 0, run(CMDS["test"], cwd=PYTHON_DIR) != 0])
        sys.exit(1 if failed else 0)
    if args.command == "clean":
        sys.exit(clean())
    if args.command in CMDS:
        sys.exit(run(CMDS[args.command], cwd=PYTHON_DIR))


if __name__ == "__main__":
    main()

File: tools/xtask/src/languages/python.rs
// python

use anyhow::Context;
use std::io::Write;
use std::path::PathBuf;

use super::{ci, run_fix_step, run_formatter};
use crate::{paths, run};

fn py_dir() -> PathBuf {
    paths::adapters_dir().join("python")
}

pub(super) fn python_format_check() -> anyhow::Result<()> {
    run_fix_step(
        "Python",
        &py_dir(),
        "uv",
        &["run", "ruff", "format", "--check", "."],
        &["run", "ruff", "format", "."],
    )
}

pub(super) fn python_format() -> anyhow::Result<()> {
    run_formatter("Python", &py_dir(), "uv", &["run", "ruff", "format", "."])
}

pub(super) fn python_lint() -> anyhow::Result<()> {
    let dir = py_dir();
    let out = run::run_capture(&dir, "uv", ["run", "ruff", "check", "."])?;
    if out.status.success() {
        return Ok(());
    }
    std::io::stdout().write_all(&out.stdout)?;
    std::io::stderr().write_all(&out.stderr)?;
    if !ci() {
        run::run(&dir, "uv", ["run", "ruff", "check", ".", "--fix"])?;
        println!("[WARN] Python lint issues auto-fixed. Stage changes before committing.");
    }
    anyhow::bail!("python lint failed")
}

pub(super) fn python_test() -> anyhow::Result<()> {
    run::run(&py_dir(), "uv", ["run", "pytest"])
}

pub(super) fn python_setup() -> anyhow::Result<()> {
    run::run(
        &py_dir(),
        "uv",
        ["sync", "--extra", "dev", "--extra", "websocket"],
    )
}

pub(super) fn python_clean() -> anyhow::Result<()> {
    let dir = py_dir();
    for name in [".venv", ".ruff_cache", ".pytest_cache"] {
        let p = dir.join(name);
        if p.is_dir() {
            std::fs::remove_dir_all(&p).with_context(|| format!("rm {}", p.display()))?;
        }
    }
    remove_tree_matching(&dir, |p| {
        let name = p.file_name().and_then(|s| s.to_str()).unwrap_or_default();
        name == "__pycache__" || name.ends_with(".egg-info") || name.ends_with(".pyc")
    })
}

fn remove_tree_matching(
    dir: &std::path::Path,
    pred: impl Fn(&std::path::Path) -> bool,
) -> anyhow::Result<()> {
    let mut stack: Vec<PathBuf> = vec![dir.to_path_buf()];
    let mut found: Vec<PathBuf> = Vec::new();
    while let Some(p) = stack.pop() {
        if let Ok(read) = std::fs::read_dir(&p) {
            for entry in read.flatten() {
                let path = entry.path();
                if path.is_dir() {
                    stack.push(path.clone());
                }
                found.push(path);
            }
        }
    }
    for p in found.iter().rev() {
        if pred(p) {
            if p.is_dir() {
                std::fs::remove_dir_all(p).with_context(|| format!("rm {}", p.display()))?;
            } else {
                let _ = std::fs::remove_file(p);
            }
        }
    }
    Ok(())
}
